core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -471,19 +472,24 @@ private static void parseObjectOrField(ParseContext context, Mapper mapper) thro
 
     private static ObjectMapper parseObject(final ParseContext context, ObjectMapper mapper, String currentFieldName) throws IOException {
         assert currentFieldName != null;
-        context.path().add(currentFieldName);
 
         ObjectMapper update = null;
         Mapper objectMapper = getMapper(mapper, currentFieldName);
         if (objectMapper != null) {
+            context.path().add(currentFieldName);
             parseObjectOrField(context, objectMapper);
+            context.path().remove();
         } else {
-            ObjectMapper.Dynamic dynamic = dynamicOrDefault(mapper, context);
+
+            // TODO: why Strings.splitStringToArray instead of String.split?
+            final String[] paths = Strings.splitStringToArray(currentFieldName, '.');
+            currentFieldName = paths[paths.length - 1];
+            Tuple<Integer, ObjectMapper> parentMapperTuple = getDynamicParentMapper(context, paths, mapper);
```

Comment:
Do we really need a tuple? Shouldn't it always be `paths.length - 1`, ie anymore more than one path piece means we had to get that many extra parent mappers? Alternatively, I had also considered adding a serialized form, ie do a join of the first `paths.length - 1` elements of path and add those as "one thing", so you are always just removing a single element at the end. I think this would work since all path serialization does is add dots between elements.